pkg/crd/cloudprivateipconfig/v1/types.go

```diff
@@ -0,0 +1,62 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +resource:path=cloudprivateipconfig
+// +kubebuilder:resource:shortName=cpip,scope=Cluster
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:printcolumn:name="Node Request",type=string,JSONPath=".spec.items[*].node"
+// +kubebuilder:printcolumn:name="IP Request",type=string,JSONPath=".spec.items[*].ip"
+// +kubebuilder:printcolumn:name="Node Assignment",type=string,JSONPath=".status.items[*].node"
+// +kubebuilder:printcolumn:name="IP Assignment",type=string,JSONPath=".status.items[*].ip"
+// CloudPrivateIPConfig is a CRD allowing the user to assign private
+// IP addresses to the primary NIC on cloud VMs. This is done by
+// specifying the spec with the Kubernetes node(s) the request
+// is meant for, as well as the IP(s) requested for that/those
+// node(s).
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired private IP request(s).
+	Spec CloudPrivateIPConfigSpec `json:"spec"`
+	// Observed status of CloudPrivateIPConfig. Read-only.
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty"`
+}
+
+type CloudPrivateIPConfigSpec struct {
+	// The list of per node requested private IPs and their corresponding node assignment.
+	Items []CloudPrivateIPConfigItem `json:"items"`
+}
+
+type CloudPrivateIPConfigStatus struct {
+	// The list of per node status of private IP assignments
+	// executed. In case an assignment fails: the item will be omitted from the
+	// status.items. I.e: if all went well the `.spec` should
+	// equal `.status`
+	Items []CloudPrivateIPConfigItem `json:"items"`
```

Comment:
would  the status be something more than the list of items as in the `Spec`? I would imagine a struct that has conditions and/status such as `Assignment Successful`, `Assignment requested`, `Assignment Pending` etc might perhaps make more sense?

Reply:
actually yea, else we would just be duplicating what is there in the request item fields right? Having status fields would also align with what the field intends to do.
maybe:
```
status: [
  {
    assignment: <nodename>-<ip-address>
    status: success
  },
  {
    assignment: <nodename>-<ip-address>
    status: failure
  },
]
```

Reply:
I am not sure it's needed, in the enhancement proposal I wrote: 

```
In case an assignment is unsuccessful: the `.status.assignment[?]` corresponding
to that `.spec.requests[?]` will not be set. This is done as to explicitly
notify the network plugin of which (if not all) requests have failed by still
issuing an UPDATE on the object.
```

I don't really see the point of a text valued status, since the network plugin will not proceed to finalizing its setup until the CNCC finishes its, and the network plugins don't have a text valued status for their egress IP assignment.
```
+--------------+                                   +-------------+                         +--------------+
|              |                                   |             |                         |              |
|              |                                   |             |                         |              |
|              |         CPIPC - Request           |             |                         |              |
|              |   +-----------+------------>      |             |                         |              |
|              |                                   |             |                         |              |
|              |                                   |             |        Cloud call       |              |
|              |                                   |             |  +--------------------> |              |
|    Network   |                                   |     CNCC    |                         |              |
|    Plugin    |                                   |             |                         |      Cloud   |
|              |                                   |             |                         |              |
|              |                                   |             |                         |              |
|              |                                   |             | Wait for cloud response                |
|              |                                   |             |   <-------------------+ |              |
|              |                                   |             |                         |              |
|              |                                   |             |                         |              |
|              |       CPIPC - status update       |             |                         |              |
|              |                                   |             |                         |              |
+--------------+    <-----------------------+      +-------------+                         +--------------+
```
I am fine with the network plugin just notifying the user that the CNCC couldn't finalize the setup.

Reply:
this is mostly for debuggability as an end-user. imagine a flow where the CNCC tried to do its thing and failed. Of course the network plugin won't proceed, but what is the indication to the end-user as to what went wrong?

Reply:
Well, if we've decided that this component is not user-facing, then there wouldn't be much more information needed for the user to debug. All they care about is if egress IP works or not, and if it doesn't: create a ticket to us and we would know to look in the CNCC.